homeassistant/components/risco/__init__.py

```diff
@@ -18,17 +31,94 @@
 from homeassistant.helpers.storage import Store
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
-from .const import DATA_COORDINATOR, DEFAULT_SCAN_INTERVAL, DOMAIN, EVENTS_COORDINATOR
+from .const import (
+    DATA_COORDINATOR,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    EVENTS_COORDINATOR,
+    TYPE_LOCAL,
+)
 
 PLATFORMS = [Platform.ALARM_CONTROL_PANEL, Platform.BINARY_SENSOR, Platform.SENSOR]
-UNDO_UPDATE_LISTENER = "undo_update_listener"
 LAST_EVENT_STORAGE_VERSION = 1
 LAST_EVENT_TIMESTAMP_KEY = "last_event_timestamp"
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass
+class LocalData:
+    """A data class for local data passed to the platforms."""
+
+    system: RiscoLocal
+    zone_updates: dict[int, Callable[[], Any]] = field(default_factory=dict)
+    partition_updates: dict[int, Callable[[], Any]] = field(default_factory=dict)
+
+
+def is_local(entry: ConfigEntry) -> bool:
+    """Return whether the entry represents an instance with local communication."""
+    return entry.data.get(CONF_TYPE) == TYPE_LOCAL
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Risco from a config entry."""
+    if is_local(entry):
+        return await _async_setup_local_entry(hass, entry)
+
+    return await _async_setup_cloud_entry(hass, entry)
+
+
+async def _async_setup_local_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    data = entry.data
+    risco = RiscoLocal(data[CONF_HOST], data[CONF_PORT], data[CONF_PIN])
+
+    try:
+        await risco.connect()
+    except CannotConnectError as error:
+        raise ConfigEntryNotReady() from error
+    except UnauthorizedError:
+        _LOGGER.exception("Failed to login to Risco cloud")
+        return False
+
+    async def _error(error: Exception) -> None:
+        _LOGGER.error("Error in Risco library: %s", error)
+
+    entry.async_on_unload(risco.add_error_handler(_error))
+
+    async def _default(command: str, result: str, *params: list[str]):
+        _LOGGER.debug(
+            "Unhandled update from Risco library: %s, %s, %s", command, result, params
+        )
+
+    entry.async_on_unload(risco.add_default_handler(_default))
+
+    local_data = LocalData(risco)
+
+    async def _zone(zone_id: int, zone: Zone):
```

Comment:
Don't forget the return value typing.
```suggestion
    async def _zone(zone_id: int, zone: Zone) -> None:
```